grpc-sys/grpc_wrap.cc

```diff
@@ -845,4 +848,133 @@ grpcwrap_ssl_server_credentials_create(
   return creds;
 }
 
+using grpc::CompletionQueue;
+
+/** Wrapper of `grpc::CompletionQueue`.
+ */
+typedef struct grpcwrap_completion_queue {
+  CompletionQueue* cq;
+} grpcwrap_completion_queue;
+
+/** Create a `grpcwrap_completion_queue`, a wrapper of `grpc_completion_queue`.
+ *
+ * Note that `grpc::CompletionQueue` wants to take the ownership of cq,
+ * but in our usage, its lifetime is maintained by `CompletionQueueHandle`
+ * in Rust.
+ */
+GPR_EXPORT grpcwrap_completion_queue* GPR_CALLTYPE
+grpcwrap_completion_queue_shadow_create(grpc_completion_queue* cq) {
+  void* ptr = gpr_malloc(sizeof(CompletionQueue));
+  CompletionQueue* shadow_cq = new (ptr) CompletionQueue(cq);
+
+  grpcwrap_completion_queue* cq_wrapper =
+      (grpcwrap_completion_queue*)gpr_malloc(sizeof(grpcwrap_completion_queue));
+  cq_wrapper->cq = shadow_cq;
+
+  return cq_wrapper;
+}
+
+/** Destroy a `grpcwrap_completion_queue`.
+ */
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_completion_queue_shadow_destroy(grpcwrap_completion_queue* cq_wrapper) {
+  // Free `grpc::CompletionQueue` without its destructor.
+  gpr_free(cq_wrapper->cq);
+  gpr_free(cq_wrapper);
+  return;
+}
+
+using grpc::Alarm;
+
+typedef Alarm grpcwrap_alarm;
+
+/** Create a completion queue alarm instance */
+GPR_EXPORT grpcwrap_alarm* GPR_CALLTYPE grpcwrap_alarm_create(void) {
+  void* ptr = gpr_malloc(sizeof(Alarm));
+  auto alarm = new (ptr) Alarm();
+  return alarm;
+}
+
+/** Set a completion queue alarm instance associated to a cq.
+ *
+ * Once the alarm notifies (see a grpcwrap_alarm_cancel) or it's destroy (see a
+ * grpcwrap_alarm_destroy), an event with a tag will be added to a cq. If the
+ * alarm notified, the event's success bit will be true, false otherwise. */
+GPR_EXPORT void GPR_CALLTYPE grpcwrap_alarm_set(grpcwrap_alarm* alarm,
+                                                grpcwrap_completion_queue* cq,
+                                                void* tag) {
+  gpr_timespec inf = gpr_inf_future(GPR_CLOCK_REALTIME);
+  alarm->Set(cq->cq, inf, tag);
+  return;
+}
+
+/** Notify a completion queue alarm. Calling this function over an alarm that
+ * has already fired has no effect. */
+GPR_EXPORT void GPR_CALLTYPE grpcwrap_alarm_cancel(grpcwrap_alarm* alarm) {
+  alarm->Cancel();
+  return;
+}
+
+/** Destroy the given completion queue alarm, cancelling it in the process. */
+GPR_EXPORT void GPR_CALLTYPE grpcwrap_alarm_destroy(grpcwrap_alarm* alarm) {
+  alarm->~Alarm();
+  gpr_free(alarm);
+  return;
+}
+
+using grpc::internal::CompletionQueueTag;
+
+/** A helper class that implements `CompletionQueueTag`.
+ *
+ * TODO: Remove it. The `grpc::Alarm` submit itself as a tag to a
```

Comment:
Is this for later work or is it for this PR?

Reply:
For later work, It's a workaround temporarily.

Reply:
Can you explain why do we need to wrap the tag like this?